### ISSUE TITLE
508-defect-1, sitewide [FOCUS]: Mobile menu does not trap focus

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#25659]
# [508-defect-1](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-2)

## Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

## Point of Contact

**VFS Point of Contact:** Josh

## User Story or Problem Statement

- As a user with low or no vision, I rely on screen reader announcements to help communicate the shift in interaction mode.
- As a user with low vision and a magnified display, focusing outside of the bounds of the menu is confusing and disorienting.
- As a user who navigates solely via keyboard, I could tab out of the menu and get lost on the underlying page or view trying to get back into the menu.

## Details

Focus is _not_ trapped within the mobile menu. For keyboard users on smaller device width sizes, this can pose an issue because the expanded mobile menu _looks_ like a modal menu that covers the entire page but _doesn't_ behave like one. 

Sighted keyboard users may believe that their focus has been completely lost unless they tab to a point that goes beyond the height of the menu.

Notice in this video demo the light grey box at the bottom left. The items being focused are changing because focus has run past the menu and into the actual page. That's not immediately clear because the mobile menu overlays on top of the entire page.

https://user-images.githubusercontent.com/14154792/120828868-069ab580-c52b-11eb-933e-59dac6d212ef.mov

## Acceptance Criteria

- [x] Focus is trapped within the mobile menu

## Environment

Sitewide

## Steps to Recreate

1. Go to va.gov and reduce your browser width until the menu collapses. 
2. Use the tab tool to open the menu 
3. Tab through and confirm that focus goes past the menu

**Comment by Rleahy22** [comment]
This should go out with the 8/31 deployment. Every page should have a unique mega menu for desktop vs mobile, and tab focus should follow the logical order.

**Comment by brianalloyd** [comment]
@joshkimux looks good to me, could you validate and close in support.

**Comment by joshkimux** [comment]
@Rleahy22 Awesome! The tab order looks great. Is there any way we might be able to trap focus within the mobile menu so the user has to select one of the options or choose to close the menu? The issue of the yellow focus halo being hidden by the mobile menu still exists 🙇

**Comment by Rleahy22** [comment]
Yup, sorry about that. I'll add that now.

**Comment by Rleahy22** [comment]
@joshkimux I just merged changes to trap focus. It should go out with the 3:00est deployment today.

**Comment by joshkimux** [comment]
@Rleahy22 Will validate this afternoon. Thank you!! 💯

**Comment by joshkimux** [comment]
@Rleahy22 Awesome! Focus is trapped! 💯 

One last little detail-- it looks like focus is getting lost when the user expands and collapses the menu. As a result:
- folks don't receive feedback that they just expanded the menu and
- folks need to tab an extra time to refocus on the menu button for it to read out its new state

Ideally we're looking for behavior similar to the submenu expand/collapsible buttons inside the menu.

Here's a video demonstrating the behavior:

https://user-images.githubusercontent.com/14154792/132882462-8ee3b299-149d-4aa4-bc10-a7e128819ce9.mov

:17s shows the behavior we want. Upon activating, the user gets immediate feedback that they're in an expanded state and the focus remains on the button 😄

**Comment by Rleahy22** [comment]
> @Rleahy22 Awesome! Focus is trapped! 💯
> 
> One last little detail-- it looks like focus is getting lost when the user expands and collapses the menu. As a result:
> 
> * folks don't receive feedback that they just expanded the menu and
> * folks need to tab an extra time to refocus on the menu button for it to read out its new state
> 
> Ideally we're looking for behavior similar to the submenu expand/collapsible buttons inside the menu.
> 
> Here's a video demonstrating the behavior:
> 
>  Screen.Recording.2021-09-10.at.11.55.14.AM.mov 
> :17s shows the behavior we want. Upon activating, the user gets immediate feedback that they're in an expanded state and the focus remains on the button 😄

Good catch, we're getting closer! I'll have a fix soon.

**Comment by Rleahy22** [comment]
The fix for keeping focus and updating the voiceover text should go out with tomorrow's deployment.

**Comment by joshkimux** [comment]
Confirmed this has been fixed on staging

https://user-images.githubusercontent.com/14154792/133830566-e0390cfa-67de-48eb-91cd-73f9602a5b81.mov
